src/main/java/com/statsig/androidsdk/StatsigNetwork.kt

```diff
@@ -27,227 +26,186 @@ private val RETRY_CODES: IntArray = intArrayOf(
-            api: String,
-            endpoint: String,
-            sdkKey: String,
-            bodyString: String,
-            responseType: KClass<T>,
-            retries: Int,
-            sharedPrefs: SharedPreferences?,
-            retryAttempt: Int = 0,
-        ): T? {
-            val url = if (api.endsWith("/")) "$api$endpoint" else "$api/$endpoint"
-            val connection: HttpURLConnection =
-                URL(url).openConnection() as HttpURLConnection
-
-            connection.requestMethod = "POST"
-            connection.setRequestProperty("Content-Type", "application/json; charset=UTF-8")
-            connection.setRequestProperty("STATSIG-API-KEY", sdkKey)
-            connection.setRequestProperty("STATSIG-CLIENT-TIME", java.lang.System.currentTimeMillis().toString())
-            connection.setRequestProperty("Accept", "application/json");
-
-            try {
-                connection.outputStream.use { os ->
-                    val input: ByteArray = bodyString.toByteArray(Charsets.UTF_8)
-                    os.write(input, 0, input.size)
-                }
-                connection.outputStream.flush()
-                connection.outputStream.close()
-
-                return runBlocking {
-                    var response: T? = Gson().fromJson(
-                        InputStreamReader(connection.inputStream, Charsets.UTF_8),
-                        responseType.java
-                    )
-                    connection.inputStream.close()
-
-                    if (connection.responseCode >= 200 && connection.responseCode < 300) {
-                        response
-                    } else if (RETRY_CODES.contains(connection.responseCode)) {
-                        if (retries > 0 && retryAttempt < retries) {
-                            Thread.sleep((100.0).pow(retryAttempt + 1).toLong())
-                            postRequestAsync(
-                                api,
-                                endpoint,
-                                sdkKey,
-                                bodyString,
-                                responseType,
-                                retries,
-                                sharedPrefs,
-                                retryAttempt + 1
-                            )
-                        } else if (sharedPrefs != null && endpoint.equals(LOGGING_ENDPOINT)) {
-                            addFailedLogRequest(sharedPrefs, bodyString)
-                            null
+    // Bug with Kotlin where any function that throws an IOException still triggers this lint warning
+    // https://youtrack.jetbrains.com/issue/KTIJ-838
+    @Suppress("BlockingMethodInNonBlockingContext")
+    private suspend inline fun <reified T : Any> postRequest(api: String, endpoint: String, sdkKey: String, bodyString: String, retries: Int, sharedPrefs: SharedPreferences?): T? {
+        return withContext(Dispatchers.IO) { // Perform network calls in IO thread
```

Comment:
Never assume the thread of a suspend function. Explicitly declare it where needed.